docs/benchmarks-design.md

```diff
@@ -0,0 +1,324 @@
+a)	ModelBase is the base class for all the public E2E models. It defines the abstract interfaces that need to be implemented by the subclasses. It has three subclasses, including PytorchModel, TFModel and ONNXModel. Each subclass will realize part of the abstract interfaces that is common for all models, such as “create_optimizer()”, “init_dataloader()”.
+b)	MicrobenchmarkBase is the base class for all the micro benchmarks. It defines the abstract interfaces that need to be implemented by the subclasses, such as “benchmarking()”, “format_output()” and so on.
+c)	DockerBase is the base class for real workloads based on docker. It also defines the abstract interfaces that need to be implemented by the subclasses, such as “prepare_image()”, “benchmarking()” and so on.
+2.	Derived classes for all implemented benchmarks, realized all the abstract interfaces. The benchmarks will be registered into Benchmark Registry.
+3.	Benchmark Registry: provides a way of benchmark registration, maintains all the registered benchmarks and supports benchmark selection by tag, which can be used to select desired benchmark, e.g., use the type of accelerator as the tag.
+The Executor on the uppermost layer is the entrance for all the benchmarks, it fetches the benchmark by name from Benchmark Registry, instantiates the benchmark with user configs and launches the benchmark. 
+Util provides some utility functions or classes such as RandomDataset generator, accelerator detector.
+
+![Structure of Benchmark Package](./assets/benchmark_package.png)
+*Figure 1 Structure of Benchmarks package*
+
+## Detailed Technical Design
+
+This chapter will describe the design details of all the components in Benchmarks package.
+
+### ModelBase Benchmarks
+
+The ModelBase benchmarks have 3-layers Inheritance Relationship. 
+
+#### Training
+
+A complete process of the general training is:
+
+Init_distributed_setting -> create_dataset -> create_dataloader -> create_model -> add_optimizer -> train
+
+The responsibility for function implementation of every layer is as Figure2. The “train” function of ModelBase will execute these functions according to the sequence in the figure. The functions that exist in derived class and not in base class are abstract functions.
+
+![Training process and the responsibility for function implementation of every layer](./assets/training-process.png)
+*Figure 2 Training process and the responsibility for function implementation of every layer*
+
+#### Inference
+
+A complete process of the general inference is:
+
+Init_distributed_setting -> create_dataset -> create_dataloader -> create_model -> inference
+
+Compared with training, it just gets rid of “add_optimizer()” operation. And the responsibility for function implementation of every layer is as Figure 3. The “inference” function of ModelBase will execute these functions according to the sequence in the figure. The functions that exist in derived class and not in base class are abstract functions.
```

Comment:
What's the meaning of `The functions that exist in derived class and not in base class are abstract functions.`?